mlrun/utils/vault.py

```diff
@@ -0,0 +1,269 @@
+
+vault_default_prefix = "v1/secret/data"
+vault_url_env_var = "MLRUN_VAULT_URL"
+vault_token_env_var = "MLRUN_VAULT_TOKEN"
+vault_role_env_var = "MLRUN_VAULT_ROLE"
+
+
+class VaultStore:
+    def __init__(self, token=None):
+        self._token = token
+        self.url = mlconf.secret_stores.vault.url or os.environ.get(vault_url_env_var)
+
+    @property
+    def token(self):
+        if not self._token:
+            self._login()
+
+        return self._token
+
+    def _login(self):
+        if self._token:
+            return
+
+        self._token = os.environ.get(vault_token_env_var)
+        if self._token:
+            return
+
+        vault_role = None
+        role_env = os.environ.get(vault_role_env_var)
+        if role_env:
+            role_type, role_val = role_env.split(":", 1)
+            vault_role = "mlrun-role-{}-{}".format(role_type, role_val)
+
+        self._login_with_jwt_token(vault_role)
+        if self._token is None:
+            logger.warning(
+                "Vault login: no vault token is available. No secrets will be accessible"
+            )
+
+    @staticmethod
+    def _generate_vault_path(
+        prefix=vault_default_prefix,
+        user=None,
+        project=None,
+        user_prefix="users",
+        project_prefix="projects",
+    ):
+        full_path = prefix + "/mlrun/{}/{}"
+        if user:
+            return full_path.format(user_prefix, user)
+        elif project:
+            return full_path.format(project_prefix, project)
+        else:
+            raise ValueError(
+                "To generate a vault secret path, either user or project must be specified"
+            )
+
+    @staticmethod
+    def _read_jwt_token():
+        # if for some reason the path to the token is not in conf, then attempt to get the SA token (works on k8s pods)
+        token_path = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+        if mlconf.secret_stores.vault.token_path:
+            # Override the default SA token in case a specific token is installed in the mlconf-specified path
+            secret_token_path = expanduser(
+                mlconf.secret_stores.vault.token_path + "/token"
+            )
+            if os.path.isfile(secret_token_path):
+                token_path = secret_token_path
+
+        with open(token_path, "r") as token_file:
+            jwt_token = token_file.read()
+
+        return jwt_token
+
+    def _vault_api_call(self, method, url, data=None):
+        self._login()
+
+        headers = {"X-Vault-Token": self._token}
+        full_url = self.url + "/" + url
+        if data:
+            data = json.dumps(data)
+
+        response = requests.request(method, full_url, headers=headers, data=data)
+
+        if not response:
+            logger.error(
+                "Vault failed the API call. Response code: ({}) - {}".format(
+                    response.status_code, response.reason
+                )
+            )
```

Comment:
Actually, you're wrong here. Quoting from [here](https://realpython.com/python-requests/):

> requests goes one step further in simplifying this process for you. If you use a Response instance in a conditional expression, it will evaluate to True if the status code was between 200 and 400, and False otherwise.

It's sort of like testing `response.ok`, only shorter.

Reply:
Walla, wasn't aware of this magic, cool
I would still say I prefer being explicit and predictable instead of counting on this magic, but that's a matter of opinion, marking as resolved